hledger-lib/Hledger/Data/AccountBalance.hs

```diff
@@ -0,0 +1,159 @@
+import Hledger.Data.Amount
+import Hledger.Data.Types
+
+
+instance Show a => Show (AccountBalances a) where
+  showsPrec d (AccountBalances h ds) =
+    showParen (d > 10) $
+        showString "AccountBalances"
+      . showString "{ abhistorical = " . shows h
+      . showString ", abdatemap = "
+      . showString "fromList " . shows (map (\(day, x) -> (ModifiedJulianDay $ toInteger day, x)) $ IM.toList ds)
+      . showChar '}'
+
+instance Foldable AccountBalances where
+  foldr f z (AccountBalances h as) = foldr f (f h z) as
+  foldl f z (AccountBalances h as) = foldl f (f z h) as
+  foldl' f z (AccountBalances h as) = let fzh = f z h in fzh `seq` foldl' f fzh as
+
+instance Foldable1 AccountBalances where
+  foldrMap1 f g (AccountBalances h as) = foldr g (f h) as
+  foldlMap1 f g (AccountBalances h as) = foldl g (f h) as
+  foldlMap1' f g (AccountBalances h as) = let fh = f h in fh `seq` foldl' g fh as
+
+instance Traversable AccountBalances where
+  traverse f (AccountBalances h as) = liftA2 AccountBalances (f h) $ traverse f as
+
+-- | The Semigroup instance for 'AccountBalance' will simply take the union of
+-- keys in the date map section. This may not be the result you want if the
+-- keys are not identical.
+instance Semigroup a => Semigroup (AccountBalances a) where
+  AccountBalances h1 as1 <> AccountBalances h2 as2 = AccountBalances (h1 <> h2) $ IM.unionWith (<>) as1 as2
+
+instance Monoid a => Monoid (AccountBalances a) where
+  mempty = AccountBalances mempty mempty
+
+-- | Construct an 'AccountBalance' from a list.
+accountBalancesFromList :: a -> [(Day, a)] -> AccountBalances a
+accountBalancesFromList h = AccountBalances h . IM.fromList . map (\(d, a) -> (fromInteger $ toModifiedJulianDay d, a))
+
+-- | Get the 'AccountBalance' associated to the period containing a given 'Day'.
+lookupAccountBalance :: Day -> AccountBalances a -> a
+lookupAccountBalance d (AccountBalances h as) =
+    maybe h snd $ IM.lookupLE (fromInteger $ toModifiedJulianDay d) as
+
+-- | Add the 'AccountBalance' to the appropriate location in 'AccountBalances'.
+insertAccountBalances :: Semigroup a => Maybe Day -> a -> AccountBalances a -> AccountBalances a
+insertAccountBalances mday b balances = case mday of
+    Nothing  -> balances{abhistorical = abhistorical balances <> b}
+    Just day -> balances{abdatemap = IM.insertWith (<>) (fromInteger $ toModifiedJulianDay day) b $ abdatemap balances}
+
+-- | Performs an operation on the contents of two 'AccountBalances'.
+--
+-- This will drop keys if they are not present in both AccountBalances.
+opAccountBalances :: (a -> b -> c) -> AccountBalances a -> AccountBalances b -> AccountBalances c
+opAccountBalances f (AccountBalances h1 as1) (AccountBalances h2 as2) =
+    AccountBalances (f h1 h2) $ IM.intersectionWith f as1 as2
+
+-- | Performs an operation on the contents of two 'AccountBalances', with
+-- separate treatments for those only in the first, only in the second, or in
+-- both 'AccountBalance's.
```

Comment:
Merges two AccountBalances, with separate treatments for balance informations only in ... ?

Reply:
Done